### ISSUE TITLE
twping - pscheduler throws error when packet interval and test interval options are given for twping test

### ISSUE THREAD

**Issue by ksandeep-palc** [perfsonar/pscheduler#1296]
1) Error seen with packet interval -

```
[admin@perf-node-2 ~]$  sudo pscheduler task rtt --count 5 --dest 192.168.122.77 **--interval PT0.1S** --source 192.168.122.15 --ip-tos 44 --ttl 64 --protocol twamp --timeout PT10S
Submitting task...
Task URL:
https://192.168.122.15/pscheduler/tasks/f665c326-e6a1-40b9-828e-7cdaeaae789d
Running with tool 'twping'
Fetching first run...

Next scheduled run:
https://192.168.122.15/pscheduler/tasks/f665c326-e6a1-40b9-828e-7cdaeaae789d/runs/34ee7688-35e8-4054-aeb2-9db60c1aa39e
Starts 2022-08-03T01:47:06-04 (~2 seconds)
Ends   2022-08-03T01:47:27-04 (~20 seconds)
Waiting for result...

Run failed.

Error:
  twping returned an error: twping: NTP: STA_NANO should be set. Make sure ntpd is running, and your NTP configuration is good.;twping: Invalid numeric value (PT0.1S) for schedule;twping: Invalid Schedule.;usage: twping [arguments] testaddr [servaddr];[arguments] are as follows: ;;   -h             print this message and exit;;              [Test Args];;   -c count       number of test packets;   -D DSCP        RFC 2474 style DSCP value for TOS byte;   -F file        save results to file;   -i wait        mean average time between packets (seconds);   -L timeout     maximum time to wait for a packet before declaring it lost (seconds);   -P portrange   port range to use during the test;   -s padding     size of the padding added to each packet (bytes);   -z delayStart  time to wait before executing test (seconds);;              [Connection Args];;   -A authmode    requested modes: [A]uthenticated, [E]ncrypted, [M]ixed, [O]pen;   -k passphrasefile     passphrasefile to use with Authenticated/Encrypted/Mixed modes;   -S srcaddr     specify the local address or interface for control connection and tests;   -B interface   specify the interface to use for control connection and tests;   -Z             do not specify IP addresses for tests packets (NAT traversal);   -u username    username to use with Authenticated/Encrypted/Mixed modes;   -4             connect using IPv4 addresses only;   -6             connect using IPv6 addresses only;;              [Output Args];;   -a alpha       report an additional percentile level for the delays;   -b bucketwidth bin size for histogram calculations;   -M             print machine (perl) readable summary;   -n units       'n','u','m', or 's';   -N count       number of test packets (to summarize per sub-session);   -Q             run the test and exit without reporting statistics;   -R             print RAW data: "SSEQ STIME SS SERR SRTIME SRS SRERR STTL RSEQ RSTIME RSS RSERR RTIME RS RERR RTTL\n";   -v[N]          print out individual delays. You can supply an optional N here to limit the print to the first N packets;   -U             Adds UNIX timestamps when printing individual delays;;Version: 4.4.4-1

Limit system diagnostics for this run:
  Hints:
    requester: 192.168.122.15
    server: 192.168.122.15
  Identified as everybody, local-interfaces
  Classified as default, friendlies
  Application: Hosts we trust to do everything
    Group 1: Limit 'always' passed
    Group 1: Want all, 1/1 passed, 0/1 failed: PASS
    Application PASSES
  Application: Defaults applied to non-friendly hosts
    Group 1: Limit 'innocuous-tests' passed
    Group 1: Limit 'throughput-default-time' failed: Test is not 'throughput'
    Group 1: Limit 'throughput-default-udp' failed: Test is not 'throughput'
    Group 1: Limit 'idleex-default' failed: Test is not 'idleex'
    Group 1: Want any, 1/4 passed, 3/4 failed: PASS
    Application PASSES
  Proposal meets limits
  Priority set at 0:
    Initial priority  (Set to 0)

Error from 192.168.122.15:
  twping returned an error: twping: NTP: STA_NANO should be set. Make sure ntpd is running, and your NTP configuration is good.;twping: Invalid numeric value (PT0.1S) for schedule;twping: Invalid Schedule.;usage: twping [arguments] testaddr [servaddr];[arguments] are as follows: ;;   -h             print this message and exit;;              [Test Args];;   -c count       number of test packets;   -D DSCP        RFC 2474 style DSCP value for TOS byte;   -F file        save results to file;   -i wait        mean average time between packets (seconds);   -L timeout     maximum time to wait for a packet before declaring it lost (seconds);   -P portrange   port range to use during the test;   -s padding     size of the padding added to each packet (bytes);   -z delayStart  time to wait before executing test (seconds);;              [Connection Args];;   -A authmode    requested modes: [A]uthenticated, [E]ncrypted, [M]ixed, [O]pen;   -k passphrasefile     passphrasefile to use with Authenticated/Encrypted/Mixed modes;   -S srcaddr     specify the local address or interface for control connection and tests;   -B interface   specify the interface to use for control connection and tests;   -Z             do not specify IP addresses for tests packets (NAT traversal);   -u username    username to use with Authenticated/Encrypted/Mixed modes;   -4             connect using IPv4 addresses only;   -6             connect using IPv6 addresses only;;              [Output Args];;   -a alpha       report an additional percentile level for the delays;   -b bucketwidth bin size for histogram calculations;   -M             print machine (perl) readable summary;   -n units       'n','u','m', or 's';   -N count       number of test packets (to summarize per sub-session);   -Q             run the test and exit without reporting statistics;   -R             print RAW data: "SSEQ STIME SS SERR SRTIME SRS SRERR STTL RSEQ RSTIME RSS RSERR RTIME RS RERR RTTL\n";   -v[N]          print out individual delays. You can supply an optional N here to limit the print to the first N packets;   -U             Adds UNIX timestamps when printing individual delays;;Version: 4.4.4-1

No further runs scheduled.
```

2) Error seen with timeout interval

```
[admin@perf-node-2 ~]$  sudo pscheduler task rtt --count 5 --dest 192.168.122.77 --source 192.168.122.15 --ip-tos 44 --ttl 64 --protocol twamp **--timeout PT10S**            
Submitting task...
Task URL:
https://192.168.122.15/pscheduler/tasks/b17ab490-9e55-425b-acb7-a6db4c1f63b0
Running with tool 'twping'
Fetching first run...

Next scheduled run:
https://192.168.122.15/pscheduler/tasks/b17ab490-9e55-425b-acb7-a6db4c1f63b0/runs/e55f6fd4-76d8-47be-a24a-03bcac1b7f58
Starts 2022-08-03T01:47:55-04 (~3 seconds)
Ends   2022-08-03T01:48:16-04 (~20 seconds)
Waiting for result...

Run failed.

Error:
  twping returned an error: twping: NTP: STA_NANO should be set. Make sure ntpd is running, and your NTP configuration is good.;twping: Invalid '-L' value. Positive float expected;usage: twping [arguments] testaddr [servaddr];[arguments] are as follows: ;;   -h             print this message and exit;;              [Test Args];;   -c count       number of test packets;   -D DSCP        RFC 2474 style DSCP value for TOS byte;   -F file        save results to file;   -i wait        mean average time between packets (seconds);   -L timeout     maximum time to wait for a packet before declaring it lost (seconds);   -P portrange   port range to use during the test;   -s padding     size of the padding added to each packet (bytes);   -z delayStart  time to wait before executing test (seconds);;              [Connection Args];;   -A authmode    requested modes: [A]uthenticated, [E]ncrypted, [M]ixed, [O]pen;   -k passphrasefile     passphrasefile to use with Authenticated/Encrypted/Mixed modes;   -S srcaddr     specify the local address or interface for control connection and tests;   -B interface   specify the interface to use for control connection and tests;   -Z             do not specify IP addresses for tests packets (NAT traversal);   -u username    username to use with Authenticated/Encrypted/Mixed modes;   -4             connect using IPv4 addresses only;   -6             connect using IPv6 addresses only;;              [Output Args];;   -a alpha       report an additional percentile level for the delays;   -b bucketwidth bin size for histogram calculations;   -M             print machine (perl) readable summary;   -n units       'n','u','m', or 's';   -N count       number of test packets (to summarize per sub-session);   -Q             run the test and exit without reporting statistics;   -R             print RAW data: "SSEQ STIME SS SERR SRTIME SRS SRERR STTL RSEQ RSTIME RSS RSERR RTIME RS RERR RTTL\n";   -v[N]          print out individual delays. You can supply an optional N here to limit the print to the first N packets;   -U             Adds UNIX timestamps when printing individual delays;;Version: 4.4.4-1

Limit system diagnostics for this run:
  Hints:
    requester: 192.168.122.15
    server: 192.168.122.15
  Identified as everybody, local-interfaces
  Classified as default, friendlies
  Application: Hosts we trust to do everything
    Group 1: Limit 'always' passed
    Group 1: Want all, 1/1 passed, 0/1 failed: PASS
    Application PASSES
  Application: Defaults applied to non-friendly hosts
    Group 1: Limit 'innocuous-tests' passed
    Group 1: Limit 'throughput-default-time' failed: Test is not 'throughput'
    Group 1: Limit 'throughput-default-udp' failed: Test is not 'throughput'
    Group 1: Limit 'idleex-default' failed: Test is not 'idleex'
    Group 1: Want any, 1/4 passed, 3/4 failed: PASS
    Application PASSES
  Proposal meets limits
  Priority set at 0:
    Initial priority  (Set to 0)

Error from 192.168.122.15:
  twping returned an error: twping: NTP: STA_NANO should be set. Make sure ntpd is running, and your NTP configuration is good.;twping: Invalid '-L' value. Positive float expected;usage: twping [arguments] testaddr [servaddr];[arguments] are as follows: ;;   -h             print this message and exit;;              [Test Args];;   -c count       number of test packets;   -D DSCP        RFC 2474 style DSCP value for TOS byte;   -F file        save results to file;   -i wait        mean average time between packets (seconds);   -L timeout     maximum time to wait for a packet before declaring it lost (seconds);   -P portrange   port range to use during the test;   -s padding     size of the padding added to each packet (bytes);   -z delayStart  time to wait before executing test (seconds);;              [Connection Args];;   -A authmode    requested modes: [A]uthenticated, [E]ncrypted, [M]ixed, [O]pen;   -k passphrasefile     passphrasefile to use with Authenticated/Encrypted/Mixed modes;   -S srcaddr     specify the local address or interface for control connection and tests;   -B interface   specify the interface to use for control connection and tests;   -Z             do not specify IP addresses for tests packets (NAT traversal);   -u username    username to use with Authenticated/Encrypted/Mixed modes;   -4             connect using IPv4 addresses only;   -6             connect using IPv6 addresses only;;              [Output Args];;   -a alpha       report an additional percentile level for the delays;   -b bucketwidth bin size for histogram calculations;   -M             print machine (perl) readable summary;   -n units       'n','u','m', or 's';   -N count       number of test packets (to summarize per sub-session);   -Q             run the test and exit without reporting statistics;   -R             print RAW data: "SSEQ STIME SS SERR SRTIME SRS SRERR STTL RSEQ RSTIME RSS RSERR RTIME RS RERR RTTL\n";   -v[N]          print out individual delays. You can supply an optional N here to limit the print to the first N packets;   -U             Adds UNIX timestamps when printing individual delays;;Version: 4.4.4-1

No further runs scheduled.
```

3) Works fine without packet interval or timeout interval

```
[admin@perf-node-2 ~]$  sudo pscheduler task rtt --count 5 --dest 192.168.122.77 --source 192.168.122.15 --ip-tos 44 --ttl 64 --protocol twamp
Submitting task...
Task URL:
https://192.168.122.15/pscheduler/tasks/6bdde187-4b9d-40da-af9e-50017339a1b3
Running with tool 'twping'
Fetching first run...

Next scheduled run:
https://192.168.122.15/pscheduler/tasks/6bdde187-4b9d-40da-af9e-50017339a1b3/runs/7c7183fc-205a-40ab-8316-ab00b092c246
Starts 2022-08-03T01:48:52-04 (~3 seconds)
Ends   2022-08-03T01:49:13-04 (~20 seconds)
Waiting for result...

1       perf-node-1 (192.168.122.77)  41 Bytes  TTL 255  RTT   0.7710 ms
2       perf-node-1 (192.168.122.77)  41 Bytes  TTL 255  RTT   0.4910 ms
3       perf-node-1 (192.168.122.77)  41 Bytes  TTL 255  RTT   0.6450 ms
4       perf-node-1 (192.168.122.77)  41 Bytes  TTL 255  RTT   0.4550 ms
5       perf-node-1 (192.168.122.77)  41 Bytes  TTL 255  RTT   0.4090 ms

0% Packet Loss  RTT Min/Mean/Max/StdDev = 0.409000/0.554000/0.771000/0.150000 ms

No further runs scheduled.
[admin@perf-node-2 ~]$
```

**Comment by ksandeep-palc** [comment]
Have created a PR - https://github.com/perfsonar/pscheduler/pull/1297

**Comment by mfeit-internet2** [comment]
The `twping` tool plugin is unusual in that it's capable of running two different tests, `rtt` and `latency`. 

The error you got revealed an inconsistency in the way times are handled between the two.  Our usual standard is to use an ISO8601 duration (`PT0.1S` as a string) but `latency` uses seconds (`0.1` as a numerc).  This happened very early in the development process six or more years ago and we never caught the mistake.

Anyway, the tool needs to handle both cases.  The PR you submitted would make it work for RTT but would break latency.  I'll close the PR you submitted and do a fix that can handle both cases.  (But thanks just the same for submitting it.)

**Comment by ksandeep-palc** [comment]
Thank you Mark, would the fix from your side make it to 5.0.0 release, just want to confirm.

**Comment by mfeit-internet2** [comment]
Yes, it will be released with 5.0.0.  It will also be in the next cut of the beta.  Watch for the `pscheduler-tool-twping` package to have a build date later than today.
